### ISSUE TITLE
qTox silently fails to update if an update is available

### ISSUE THREAD

**Issue by LittleVulpix** [qTox/qTox#3910]
##### Brief Description

OS: Windows 10 x64 AU, latest. 
qTox version:1.4.1-1.6.0   
Commit hash:  https://github.com/qTox/qTox/commit/5648106b34fc7dc3ad4e169651a0cbbd0139170a

Reproducible: Always

##### Steps to reproduce

1. Set qTox to update automatically
2. Have non-latest version of qTox
3. Start it

##### Observed Behavior
Nothing happens

##### Expected Behavior
You are prompted to update

##### Additional Info
Log1:

```
[01:17:38.360 UTC] src/persistence/profile.cpp:109 : Debug: Loading tox save  "C:/Users/Vulpix/AppData/Roaming/tox\\toxmeio.tox"
[01:17:38.360 UTC] src/persistence/settings.cpp:404 : Debug: "Saving global settings at C:/Users/Vulpix/AppData/Roaming/tox\\qtox.ini"
[01:17:38.375 UTC] src/persistence/settings.cpp:319 : Debug: Loading personal settings from "C:/Users/Vulpix/AppData/Roaming/tox/toxmeio.ini"
[01:17:38.375 UTC] src/nexus.cpp:84 : Debug: Starting up
[01:17:39.679 UTC] src/core/core.cpp:238 : Debug: Loading user profile
[01:17:39.680 UTC] src/persistence/profile.cpp:290 : Debug: Loading tox save  "C:/Users/Vulpix/AppData/Roaming/tox\\toxmeio.tox"
[01:17:39.680 UTC] src/core/core.cpp:133 : Warning: Core starting with IPv6 disabled. LAN discovery may not work properly.
[01:17:39.686 UTC] src/widget/systemtrayicon.cpp:90 : Debug: Using the Qt backend
[01:17:39.687 UTC] src/core/core.cpp:307 : Debug: Self avatar not found, will broadcast empty avatar to friends
[01:17:40.889 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[01:17:41.502 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[01:17:41.952 UTC] src/net/autoupdate.cpp:225 : Debug: timestamp: 1478912404 , str: "v1.5.2-8-gfd04a28"
[01:17:41.953 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[01:17:42.414 UTC] src/net/autoupdate.cpp:433 : Debug: Need to update 23 files
[01:17:42.416 UTC] src/net/autoupdate.cpp:486 : Debug: "Downloading '\u000E./libeay32.dl' ..."
[01:17:42.418 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[01:17:42.486 UTC] src/core/core.cpp:415 : Debug: "Connecting to 194.249.212.109:33445 (fluke571)"
[01:17:42.487 UTC] src/core/core.cpp:415 : Debug: "Connecting to 185.25.116.107:33445 (MAH69K)"
[01:17:42.855 UTC] src/net/autoupdate.cpp:391 : Warning: "getUpdateFile: network error: Error downloading https://qtox-win.pkg.tox.chat/qtox/win64/files/8d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c - server replied: Not Found"
[01:17:42.855 UTC] src/net/autoupdate.cpp:499 : Critical: downloadUpdate: Error downloading a file, aborting...
[01:17:48.111 UTC] src/core/core.cpp:415 : Debug: "Connecting to 144.76.60.215:33445 (sonOfRa)"
[01:17:48.112 UTC] src/core/core.cpp:415 : Debug: "Connecting to 23.226.230.47:33445 (stal)"
[01:17:53.738 UTC] src/core/core.cpp:415 : Debug: "Connecting to 192.99.168.140:33445 (WIeschie)"
[01:17:53.739 UTC] src/core/core.cpp:415 : Debug: "Connecting to 198.46.138.44:33445 (nurupo)"
[01:17:56.181 UTC] src/core/core.cpp:379 : Debug: Connected to the DHT
```

Log2:
```
[19:07:29.657 UTC] persistence/profile.cpp:117 : Debug: Loading tox save  "C:/Users/Vulpix/AppData/Roaming/tox\\toxmeio.tox"
[19:07:29.673 UTC] persistence/settings.cpp:494 : Debug: "Saving global settings at C:/Users/Vulpix/AppData/Roaming/tox\\qtox.ini"
[19:07:29.704 UTC] persistence/settings.cpp:358 : Debug: Loading personal settings from "C:/Users/Vulpix/AppData/Roaming/tox/toxmeio.ini"
[19:07:29.704 UTC] nexus.cpp:85 : Debug: Starting up
[19:07:30.789 UTC] core/core.cpp:247 : Debug: Loading user profile
[19:07:30.790 UTC] persistence/profile.cpp:312 : Debug: Loading tox save  "C:/Users/Vulpix/AppData/Roaming/tox\\toxmeio.tox"
[19:07:30.790 UTC] core/core.cpp:133 : Warning: Core starting with IPv6 disabled. LAN discovery may not work properly.
[19:07:30.795 UTC] core/core.cpp:316 : Debug: Self avatar not found, will broadcast empty avatar to friends
[19:07:30.814 UTC] widget/systemtrayicon.cpp:90 : Debug: Using the Qt backend
[19:07:32.422 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[19:07:32.638 UTC] core/core.cpp:424 : Debug: "Connecting to 148.251.23.146:2306 (pucetox)"
[19:07:32.639 UTC] core/core.cpp:424 : Debug: "Connecting to 108.61.165.198:33445 (ray65536)"
[19:07:32.949 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[19:07:33.426 UTC] net/autoupdate.cpp:225 : Debug: timestamp: 1479582004 , str: "v1.6.0"
[19:07:33.426 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[19:07:33.897 UTC] net/autoupdate.cpp:433 : Debug: Need to update 23 files
[19:07:33.899 UTC] net/autoupdate.cpp:486 : Debug: "Downloading '\u000E./libeay32.dl' ..."
[19:07:33.901 UTC] :0 : Warning: QObject::connect: Cannot connect (null)::stateChanged(QNetworkSession::State) to QNetworkReplyHttpImpl::_q_networkSessionStateChanged(QNetworkSession::State)
[19:07:34.356 UTC] net/autoupdate.cpp:391 : Warning: "getUpdateFile: network error: Error downloading https://qtox-win.pkg.tox.chat/qtox/win64/files/8d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c - server replied: Not Found"
[19:07:34.356 UTC] net/autoupdate.cpp:499 : Critical: downloadUpdate: Error downloading a file, aborting...
[19:07:38.263 UTC] core/core.cpp:424 : Debug: "Connecting to 2001:19f0:5000:8121:5054:ff:fe1c:9ded:33445 (ray65536)"
[19:07:38.263 UTC] core/core.cpp:424 : Debug: "Connecting to srv1.ricin.im:1337 (Ricin.im)"
[19:07:42.106 UTC] core/core.cpp:388 : Debug: Connected to the DHT

```
The file really doesn't exist on the server, so.. not sure what's going on there. I updated manually to 1.6.0 and now I see there is a new version again and I still didn't get any prompt :p

**Comment by nurupo** [comment]
Looks like 

>net/autoupdate.cpp:391 : Warning: "getUpdateFile: network error: Error downloading https://qtox-win.pkg.tox.chat/qtox/win64/files/8d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c - server replied: Not Found"

is the culprit and it looks like an off-by-one error, as there exists a file with very similar name https://qtox-win.pkg.tox.chat/qtox/win64/files/d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c5 on the web server.

Here is the comparison of the filenames in the monospace font:

```
# qTox is trying to download an non-existing file
https://qtox-win.pkg.tox.chat/qtox/win64/files/8d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c
# The actual file that exists on the server
https://qtox-win.pkg.tox.chat/qtox/win64/files/d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c5
````

Notice that the actual file that exists on the server doesn't have `8` at the beginning of its name and it has an additional `5` at the end.

If we were to look into https://qtox-win.pkg.tox.chat/qtox/win64/flist , we would find the next string in it

```
8d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c5
```

Which is exactly the file name of the existing on the webserver file, if we were to ignore the first `8`.

In fact, that's true for all other strings in that `flist` file, e.g. there is 

```
8e653653348f8548a71c61047f8b9970e4e58e7f63df49de9a0698087
```

and there exists a file on the web server with the same name but without `8` at the beginning: https://qtox-win.pkg.tox.chat/qtox/win64/files/e653653348f8548a71c61047f8b9970e4e58e7f63df49de9a0698087

So this looks like an off-by-one error, when qTox read filename 1 byte too early, i.e. it didn't skip the first `8` and read the fixed number of bytes, which resulted in truncating the `5` at the end in the original error

>net/autoupdate.cpp:391 : Warning: "getUpdateFile: network error: Error downloading https://qtox-win.pkg.tox.chat/qtox/win64/files/8d2d849d6954588eff3f5e19b3130451a40cb92f033ff840e2d8b62c - server replied: Not Found"

It's also possible for this to be an error in the server-side code @tux3 is using to generate the `flist` file and the files under the `files/` directory.

**Comment by zetok** [comment]
Update: there's nothing that can be done about the current qTox instances in the wild.

Starting from qTox 1.7 updater will work. :/

**Comment by LittleVulpix** [comment]
So people who don't know about this will probably never update their qtox... :(

**Comment by nurupo** [comment]
Well, one thing that could be done is to keep https://qtox-win.pkg.tox.chat/qtox update repository compatible with qTox versions 1.5 and 1.6, i.e. keep it in the broken format to work around the bug introduced in 1.5 by #3380, so that 1.5 and 1.6 could update to 1.7. There also needs to be a new update repository with a fixed format somewhere, e.g. https://qtox-win.pkg.tox.chat/qtox-fixed/, which is what qTox 1.7 would be hardcoded to use, instead of the current hardcoded https://qtox-win.pkg.tox.chat/qtox/. That way qTox 1.5 and 1.6 would be able to update to 1.7, and 1.7 would be able to update further.

**Comment by zetok** [comment]
Fixed for >=1.7. Nothing can be done for the past releases, thus closing.
